cli/src/errors/error.rs

```diff
@@ -1,49 +1,92 @@
-#[cfg(backtrace)]
-use std::backtrace::Backtrace;
 use std::convert::From;
+use std::error::Error;
+use std::fmt;
 
 use anyhow::Result;
 use log::error;
-use thiserror::Error;
 
+use crate::errors::ErrorDetails;
 use crate::errors::ExitCode;
 
-#[derive(Error, Debug)]
-#[error("{msg}")]
+#[derive(Debug)]
 pub struct ApolloError {
-    msg: String,
-    #[cfg(backtrace)]
-    backtrace: Backtrace,
-    /// The result of `error.exit_code()`.
-    exit_code: ExitCode,
+    inner: Box<Inner>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    kind: ErrorDetails,
+    source: Option<Box<dyn Error>>,
 }
 
 #[allow(dead_code)]
 impl ApolloError {
     /// Returns the process exit code that should be returned if the process exits with this error.
     pub fn exit_code(&self) -> ExitCode {
-        self.exit_code
+        self.inner.kind.exit_code()
+    }
+
+    pub fn from_source<E>(source: E, kind: ErrorDetails) -> Self
+    where
+        E: Into<Box<dyn Error>>,
+    {
+        ApolloError {
+            inner: Box::new(Inner {
+                kind,
+                source: Some(source.into()),
+            }),
+        }
+    }
+
+    pub fn kind(&self) -> &ErrorDetails {
+        &self.inner.kind
     }
 }
 
-/// The failure trait for all Apollo errors.
-pub trait ApolloFail: std::error::Error {
-    /// Returns the process exit code that should be returned if the process exits with this error.
-    fn exit_code(&self) -> ExitCode;
+impl fmt::Display for ApolloError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.inner.kind.fmt(f)
+    }
 }
 
-impl<T: ApolloFail> From<T> for ApolloError {
-    fn from(error: T) -> Self {
-        let exit_code = error.exit_code();
+impl Error for ApolloError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
```

Comment:
Q: What's the purpose of using the `'static` lifetime for the response?

Reply:
The `'static` (as I understand) supports downcasting the trait object back to a concrete type. [A few](https://rust-lang.github.io/rfcs/2504-fix-error.html#why-cause---source) [reference](https://users.rust-lang.org/t/what-does-it-mean-to-return-dyn-error-static/37619/12) links.

Tbh I'm not totally sure this is right either!